apps/api/src/app/step-schemas/usecases/get-step-schema/get-step-schema.usecase.ts

```diff
@@ -1,39 +1,149 @@
-import { Injectable } from '@nestjs/common';
-import {
-  ActionStepEnum,
-  actionStepSchemas,
-  ChannelStepEnum,
-  channelStepSchemas,
-  Schema,
-  StepType,
-} from '@novu/framework';
+import { BadRequestException, Injectable } from '@nestjs/common';
+
+import { JsonSchema, Schema, StepType } from '@novu/framework';
+import { NotificationStepEntity, NotificationTemplateRepository } from '@novu/dal';
 
-import { GetStepSchemaCommand } from './get-step-schema.command';
-import { SchemaOutput } from '../../types';
+import {
+  GetStepTypeSchemaCommand,
+  GetExistingStepSchemaCommand,
+  GetStepSchemaCommand,
+} from './get-step-schema.command';
+import { StepSchemaDto } from '../../dtos/step-schema.dto';
+import { mapStepTypeToOutput, mapStepTypeToResult } from '../../types';
 
 @Injectable()
 export class GetStepSchema {
-  async execute(command: GetStepSchemaCommand): Promise<SchemaOutput> {
-    return StepTypeToOutputMap[command.stepType];
+  constructor(private readonly notificationTemplateRepository: NotificationTemplateRepository) {}
+
+  async execute(command: GetStepSchemaCommand): Promise<StepSchemaDto> {
+    if (isGetByStepType(command)) {
+      return { controls: buildControlsSchema(command.stepType), variables: buildVariablesSchema() };
+    }
+
+    if (isGetByExistingStep(command)) {
+      const { currentStep, previousSteps } = await this.findSteps(command);
+
+      return {
+        controls: buildControlsSchema(currentStep?.template?.type as StepType),
+        variables: buildVariablesSchema(previousSteps),
+      };
+    }
+
+    throw new BadRequestException('Invalid command');
+  }
+
+  private async findSteps(command: GetExistingStepSchemaCommand) {
+    const workflow = await this.notificationTemplateRepository.findByIdQuery({
+      id: command.workflowId,
+      environmentId: command.environmentId,
+    });
+
+    if (!workflow) {
+      throw new BadRequestException(`No workflow found with the given id ${command.workflowId}`);
+    }
+
+    const currentStep = workflow.steps.find((stepItem) => stepItem._id === command.stepId);
+
+    if (!currentStep) {
+      throw new BadRequestException(`No step found with the given id ${command.stepId}`);
```

Comment:
Could we create a new error type that includes all relevant context, such as the step we searched for and the actual steps received? This would provide more clarity for troubleshooting.

Reply:
we are providing the step id we searched for, what else can we pass here? there are no steps received in this situation